### ISSUE TITLE
Inncorect checksum verification (PostgreSQL)

### ISSUE THREAD

**Issue by boosterKRD** [wal-g/wal-g#1140]
### Database name
PostgreSQL 

# Issue description
Incorrect behavior when checking checksums with the -v flag (wal-g ver 1.1)

### Describe your problem
If the data is corrupted in such a way that some of the data moves to the next page, then walg will behave incorrectly when checking checksums with the -v flag. Walg will no find bad checksums (it is not able to see data corruption.).

### Please provide steps to reproduce
**--1. create test table**
drop table  public.test_table;
CREATE TABLE public.test_table (  --create table
	id int4 NOT NULL,
	name varchar
)
WITH (
	fillfactor=100
);

**--2. Add some data** 
insert into public.test_table  (id, name)   values (1, 'MIKE1');
insert into public.test_table  (id, name)   values (2, 'OLGA');
insert into public.test_table  (id, name)   values (3, 'MISHA');

**--3. Find file in pg_data**
[root@test27-db-01]# psql -U postgres -p 5435 -Atc "select format('%s/%s',  current_setting('data_directory'),  pg_relation_filepath('test_table'))"  test_db
###/data/pg_data/postgres/base/16495/16525

**--4. make checkpoint** 
[root@test27-db-01]# psql -U postgres -p 5435  -c "checkpoint" test_db

**--5. Checking for the presence of data in the file** 
[root@test27-db-01]# hexdump -C /data/pg_data/postgres/base/16495/16525

**--6. Corrupt data in file**
[root@test27-db-01]# echo -n "Hello!" | dd conv=notrunc oflag=seek_bytes seek=8189 bs=6 count=1 of=/data/pg_data/postgres/base/16495/16525

**--7. Checking for the presence of data in the file again**  
[root@test27-db-01]# hexdump -C /data/pg_data/postgres/base/16495/16525
00000000  00 00 00 00 f0 17 03 09  47 e3 00 00 24 00 88 1f  |........G...$...|
00000010  00 20 04 20 00 00 00 00  d8 9f 44 00 b0 9f 42 00  |. . ......D...B.|
00000020  88 9f 44 00 00 00 00 00  00 00 00 00 00 00 00 00  |..D.............|
00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00001f80  00 00 00 00 00 00 00 00  0f 02 00 00 00 00 00 00  |................|
00001f90  0d 00 00 00 00 00 00 00  03 00 02 00 02 08 18 00  |................|
00001fa0  03 00 00 00 0d 4d 49 53  48 41 00 00 00 00 00 00  |.....MISHA......|
00001fb0  0f 02 00 00 00 00 00 00  0c 00 00 00 00 00 00 00  |................|
00001fc0  02 00 02 00 02 08 18 00  02 00 00 00 0b 4f 4c 47  |.............OLG|
00001fd0  41 00 00 00 00 00 00 00  0f 02 00 00 00 00 00 00  |A...............|
00001fe0  0b 00 00 00 00 00 00 00  01 00 02 00 02 08 18 00  |................|
00001ff0  01 00 00 00 0d 4d 41 52  41 54 00 00 00 48 65 6c  |.....MARAT...Hel|
00002000  6c 6f 21                                          |lo!|
00002003

**--8. Restart Postgres and do attempt to access data from postgres and make backup by walg** 
[root@test27-db-01]# systemctl restart postgresql-13
[root@test27-db-01]# psql -U postgres -p 5435  -c "select * from test_table" test_db
SECCESS POstgres see data damage
_WARNING:  page verification failed, calculated checksum 24693 but expected 58183
ERROR:  invalid page in block 0 of relation base/16495/16525_

[root@test27-db-01]# su -c "/usr/local/bin/wal-g backup-push -f -s -v /data/pg_data/postgres" postgres
_FAIL: WALg don't see any data coruption (There aren't any warnings in stderr)
### Please add config and wal-g stdout/stderr logs for debug purpose
INFO: 2021/11/08 12:25:52.588586 Selecting the latest backup as the base for the current delta backup...
INFO: 2021/11/08 12:25:52.602764 Doing full backup.
INFO: 2021/11/08 12:25:52.610169 Calling pg_start_backup()
INFO: 2021/11/08 12:25:52.845646 Starting a new tar bundle
INFO: 2021/11/08 12:25:52.845704 Walking ...
INFO: 2021/11/08 12:25:52.846017 Starting part 1 ...
INFO: 2021/11/08 12:25:52.846118 Starting part 2 ...
INFO: 2021/11/08 12:25:53.081292 Packing ...
INFO: 2021/11/08 12:25:53.081896 Finished writing part 2.
INFO: 2021/11/08 12:26:25.994541 Finished writing part 1.
INFO: 2021/11/08 12:26:26.448139 Starting part 3 ...
INFO: 2021/11/08 12:26:26.448194 /global/pg_control
INFO: 2021/11/08 12:26:26.448730 Finished writing part 3.
INFO: 2021/11/08 12:26:26.449340 Calling pg_stop_backup()
INFO: 2021/11/08 12:27:49.604398 Starting part 4 ...
INFO: 2021/11/08 12:27:49.604576 backup_label
INFO: 2021/11/08 12:27:49.604591 tablespace_map
INFO: 2021/11/08 12:27:49.609611 Finished writing part 4.
INFO: 2021/11/08 12:27:52.810995 Wrote backup with name base_00000001000000000000000B_

**Comment by x4m** [comment]
Hi! Information about the count of checksum violations is added to backup sentinel.
Pull request with a warning to log would be welcome. But several checksum violations may be observed during normal backup due to torn pages, and these violations are absolutely legal. I'd recommend monitoring checksum violations at certain threshold.

**Comment by boosterKRD** [comment]
Hi, sorry for the intrusiveness. The main problem in this case that the file size base/16495/16525 eq 8195 byte and checking for  checksum violations is skipping (VerifyPagedFileBase)
```

in tar_ball_file_packer.go
func verifyFile(path string, fileInfo os.FileInfo, fileReader io.Reader, isIncremented bool) ([]uint32, error) {
	if !isPagedFile(fileInfo, path) { **//isPagedFile will be FALSE for file 16525 (info.Size()%DatabasePageSize != 0)**
		_, err := io.Copy(ioutil.Discard, fileReader)
		return nil, err
	}

	if isIncremented {
		return VerifyPagedFileIncrement(path, fileInfo, fileReader)
	}
	return VerifyPagedFileBase(path, fileInfo, fileReader)
}
```
_```
some debug logs
WARNING: 2021/11/08 18:38:57.711044 filePath  ---  /data/pg_data/postgres/base/16495/16533
WARNING: 2021/11/08 18:38:57.711049 DatabasePageSize  ---  %!s(int64=8192)
WARNING: 2021/11/08 18:38:57.711052 size ---  %!s(int64=8195)
WARNING: 2021/11/08 18:38:57.711054 mod_num  ---  %!s(int64=3)
```_

**Comment by x4m** [comment]
Yeah, a short-reading page should be treated as a clear sign of corruption. That's a bug.

Looking at this code I wonder: why do we read file one more time for verification?  Shouldn't we verify file just during upload?

**Comment by usernamedt** [comment]
> Yeah, a short-reading page should be treated as a clear sign of corruption. That's a bug.
> 
> 
> 
> Looking at this code I wonder: why do we read file one more time for verification?  Shouldn't we verify file just during upload?

Actually, we don't read the file second tine (there is TeeReader already implemented for this purpose).
